src/v/storage/mvlog/batch_collecting_stream_utils.cc

```diff
@@ -0,0 +1,88 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#include "storage/mvlog/batch_collecting_stream_utils.h"
+
+#include "base/vlog.h"
+#include "storage/mvlog/batch_collector.h"
+#include "storage/mvlog/logger.h"
+#include "storage/record_batch_utils.h"
+
+namespace storage::experimental::mvlog {
+
+std::ostream& operator<<(std::ostream& o, collect_stream_outcome out) {
+    switch (out) {
+    case collect_stream_outcome::buffer_full:
+        return o << "collect_stream_outcome::buffer_full";
+    case collect_stream_outcome::end_of_stream:
+        return o << "collect_stream_outcome::end_of_stream";
+    case collect_stream_outcome::stop:
+        return o << "collect_stream_outcome::stop";
+    }
+}
+
+ss::future<result<collect_stream_outcome, errc>>
+collect_batches_from_stream(entry_stream* entries, batch_collector* collector) {
+    while (true) {
+        auto entry_res = co_await entries->next();
+        if (entry_res.has_error()) {
+            co_return entry_res.error();
+        }
+        auto& entry = entry_res.value();
+        if (!entry.has_value()) {
+            co_return collect_stream_outcome::end_of_stream;
+        }
+        switch (entry->hdr.type) {
```

Comment:
It looks like this switch statement has only one case.

Reply:
Yeah, this will change as we add more header types. Left this as a switch for the time being, but a later PR will add more.